fortimanager_connector.py

```diff
@@ -0,0 +1,462 @@
+# File: fortimanager_connector.py
+#
+# Copyright (c) 2023 Splunk Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific language governing permissions
+# and limitations under the License.
+
+# Python 3 Compatibility imports
+from __future__ import print_function, unicode_literals
+
+import json
+import re
+import traceback
+
+# Phantom App imports
+import phantom.app as phantom
+import requests
+from bs4 import BeautifulSoup
+from phantom.action_result import ActionResult
+from phantom.base_connector import BaseConnector
+from pyFMG.fortimgr import FortiManager
+
+from fortimanager_consts import *
+
+
+class RetVal(tuple):
+
+    def __new__(cls, val1, val2=None):
+        return tuple.__new__(RetVal, (val1, val2))
+
+
+class FortimanagerConnector(BaseConnector):
+
+    def __init__(self):
+
+        # Call the BaseConnectors init first
+        super(FortimanagerConnector, self).__init__()
+
+        self._state = None
+
+        self._base_url = None
+
+        self._host = None
+        self._username = None
+        self._password = None
+
+        self._api_key = None
+
+    def _login(self, action_result):
+        if self._username and self._password:
+            fmg_instance = FortiManager(self._host, self._username, self._password, debug=True, disable_request_warnings=True)
+        elif self._api_key:
+            fmg_instance = FortiManager(self._host, apikey=self._api_key, debug=True, disable_request_warnings=True)
+        else:
+            raise Exception("The asset configuration requires either an API key or a username and password.")
+        fmg_instance.login()
+        return fmg_instance
+
+    def _process_empty_response(self, response, action_result):
+        if response.status_code == 200:
+            return RetVal(phantom.APP_SUCCESS, {})
+
+        return RetVal(
+            action_result.set_status(
+                phantom.APP_ERROR, "Empty response and no information in the header"
+            ), None
+        )
+
+    def _process_html_response(self, response, action_result):
```

Comment:
We can probably remove these process response functions unless we plan to make any REST calls.